lib/CL/devices/remote/network_discovery.c

```diff
@@ -0,0 +1,687 @@
+  POCL_MSG_PRINT_REMOTE ("Avahi Browser: Browsing started \n");
+
+  return errcode;
+
+ERROR:
+
+  clear_avahi ();
+  return errcode;
+}
+#endif
+
+/*****************************************************************************/
+#if defined(ENABLE_REMOTE_DISCOVERY_DHT)
+
+const char *common_key = NULL;
+
+struct op_context
+{
+  dht_runner *node;
+  atomic_bool stop;
+};
+
+struct listen_context
+{
+  dht_runner *node;
+  dht_op_token *token;
+  size_t count;
+};
+
+static bool
+dht_value_callback (const dht_value *value, bool expired, void *user_data)
+{
+
+  struct listen_context *ctx = (struct listen_context *)user_data;
+
+  if (expired)
+    ctx->count--;
+  else
+    ctx->count++;
+
+  dht_data_view data = dht_value_get_data (value);
+  char
+    key[55]; /* size of AVAHI_ADDRESS_STR_MAX + 15, which is max size of IP */
+             /* address + 15 */
+  char id[33]; /* server id is 32char + '\0' */
+  int dev_count = 0;
+
+  /* The device information from the server is separated using ~ as the
+   * delimiter. The client should know how to interpret/decode the server
+   * data. */
+  char *ddata;
+  char *token;
+  ddata = strndup ((char *)data.data, data.size);
+
+  token = strtok (ddata, "~");
+  snprintf (id, strlen (token) + 1, "%s", token);
+
+  token = strtok (NULL, "~");
+  snprintf (key, strlen (token) + 1, "%s", token);
+
+  while (token)
+    {
+      token = strtok (NULL, "~");
+      dev_count++;
+    }
+  POCL_MEM_FREE (ddata);
+  dev_count = dev_count / 4;
+
+  handle_server_discovery (id, common_key, key, "DHT", dev_count);
+
+  return true;
+}
+
+static void
+listen_context_free (void *user_data)
+{
+  struct listen_context *ctx = (struct listen_context *)user_data;
+  dht_op_token_delete (ctx->token);
+  POCL_MEM_FREE (ctx);
+}
+
+void
+dht_shutdown_callback (void *user_data)
+{
+  struct op_context *ctx = (struct op_context *)user_data;
+  atomic_store (&ctx->stop, true);
+}
+
+static dht_infohash
+get_hash (const char *key_str)
+{
+  dht_infohash key;
+  dht_infohash_from_hex_null (&key, key_str);
+  if (dht_infohash_is_zero (&key))
+    {
+      dht_infohash_get_from_string (&key, key_str);
+    }
+  return key;
+}
+
+/**
+ * Initialize DHT node and start the listener on the DHT network and the given
+ * key.
+ */
+static cl_int
+init_dht_discovery ()
+{
+  cl_int errcode = CL_SUCCESS;
+
+  dht_runner *node = dht_runner_new ();
+  dht_runner_config dht_config;
+  dht_runner_config_default (&dht_config);
+
+  /* Port to start the DHT taken from environment or default used.*/
+  const int port = pocl_get_int_option (POCL_REMOTE_DHT_PORT_ENV, 4222);
+  /* DHT bootstrap node taken from environment or default used.*/
+  const char *bootstrap
+    = pocl_get_string_option (POCL_REMOTE_DHT_BOOTSTRAP_ENV, NULL);
+  if (bootstrap == NULL)
+    {
+      POCL_MSG_ERR ("DHT bootstrap node environment variable not specified, "
+                    "server advertisement failed! \n");
```

Comment:
This is client code, no?

Reply:
Ah yes, I copied the comment from the server and forgot to modify.